bucket-scanner/pkg/counter/replication/counter_set.go

```diff
@@ -0,0 +1,119 @@
+package replication
+
+import (
+	"sync"
+
+	"github.com/scality/backbeat/bucket-scanner/pkg/types"
+)
+
+type (
+	CounterSet struct {
+		sync.Mutex
+		Global           *ByStatusCounterSet
+		ByTargetLocation map[types.SiteName]*ByStatusCounterSet
+
+		name string
+	}
+
+	SerializedCounterSet struct {
+		Global     *SerializedByStatusCounterSet                    `json:"global"`
+		ByLocation map[types.SiteName]*SerializedByStatusCounterSet `json:"byLocation"`
+	}
+)
+
+func New(name string) *CounterSet {
+	return &CounterSet{
+		Global:           NewByStatusCounterSet(name + ":global"),
+		ByTargetLocation: make(map[types.SiteName]*ByStatusCounterSet),
+
+		name: name,
+	}
+}
+
+func (cs *CounterSet) CopyTo(other *CounterSet) {
+	other.Lock()
+	defer other.Unlock()
+
+	cs.Global.CopyTo(other.Global)
+
+	other.ByTargetLocation = make(map[types.SiteName]*ByStatusCounterSet)
+	for k, v := range cs.ByTargetLocation {
+		n := NewByStatusCounterSet(other.name + ":" + string(k))
+		v.CopyTo(n)
+		other.ByTargetLocation[k] = n
+	}
+}
+
+func (cs *CounterSet) Add(other *CounterSet) {
+	cs.Lock()
+	defer cs.Unlock()
+
+	cs.Global.Add(other.Global)
+
+	for k, v := range other.ByTargetLocation {
+		m := cs.getTargetLocation(k)
+		m.Add(v)
+	}
+}
+
+func (cs *CounterSet) Reset() {
+	cs.Lock()
+	defer cs.Unlock()
+
+	cs.Global.Reset()
+	cs.ByTargetLocation = make(map[types.SiteName]*ByStatusCounterSet)
+}
+
+func (cs *CounterSet) GetTargetLocation(location types.SiteName) *ByStatusCounterSet {
+	cs.Lock()
+	defer cs.Unlock()
+
+	return cs.getTargetLocation(location)
+}
+
+func (cs *CounterSet) getTargetLocation(location types.SiteName) *ByStatusCounterSet {
+	bscs, ok := cs.ByTargetLocation[location]
+	if !ok {
+		bscs = NewByStatusCounterSet(cs.name + ":" + string(location))
+		cs.ByTargetLocation[location] = bscs
+	}
+
+	return bscs
+}
+
+func (cs *CounterSet) Serialize() *SerializedCounterSet {
+	cs.Lock()
+	defer cs.Unlock()
+
+	byLocation := map[types.SiteName]*SerializedByStatusCounterSet{}
+
+	for location, bscs := range cs.ByTargetLocation {
+		byLocation[location] = bscs.Serialize()
+	}
+
+	return &SerializedCounterSet{
+		Global:     cs.Global.Serialize(),
+		ByLocation: byLocation,
+	}
+}
+
+func (cs *CounterSet) LoadSerialized(s *SerializedCounterSet) {
```

Comment:
will `LoadSerialized` always be called in a safe manner?

Please correct me if I am wrong, with the lock claim/release being done per location, the result may be unexpected  if the `LoadSerialized` is called along other `CounterSet` write methods.